central/image/service/service_impl.go

```diff
@@ -457,13 +466,42 @@ func (s *serviceImpl) EnrichLocalImageInternal(ctx context.Context, request *v1.
 	// Due to discrepancies in digests retrieved from metadata pulls and k8s, only upsert if the request
 	// contained a digest. Do not upsert if a previous scan exists and there were errors with this scan
 	// since it could lead to us overriding an enriched image with a non-enriched image.
-	if imgID != "" && !(hasErrors && imgExists) {
+	// Also do not upsert if there is a request id, this enables the caller to determine how to handle
+	// the results (and also prevents multiple upserts for the same image).
+	if imgID != "" && !(hasErrors && imgExists) && request.GetRequestId() == "" {
```

Comment:
The originator of the request currently decides that, and in most cases it will be persisted.

For example, with `roxctl image scan` the call flow would look something like this:

1. `ScanImage()` API invoked, and [enrich requested](https://github.com/stackrox/stackrox/blob/f1abdaf371b14126c746530d96a2d634eec4401e/central/image/service/service_impl.go#L306)
1. Enrich eventually delegates to secured cluster and result returned via `EnrichLocalImageInternal` [here](https://github.com/stackrox/stackrox/blob/f1abdaf371b14126c746530d96a2d634eec4401e/central/image/service/service_impl.go#L482)
1. `ScanImage()` continues [persists the result](https://github.com/stackrox/stackrox/blob/f1abdaf371b14126c746530d96a2d634eec4401e/central/image/service/service_impl.go#L312-L317)

Without that check, both steps 2 and 3 would save the same image to the DB

Reply:
Also just realized we re-save cached images in these flows as well: `roxctl image scan -> image pulled from DB -> image re-saved to DB -> image returned`

There appear to be some functions (such as `FillScanStats` in `pkg/images/enricher/enricher_impl.go`) that are called on every save that modify image stats in place, without it the stats would not be returned on cache hit (the summary fields populated by `FillScanStats` do not appear to be copied into the image in place when pulled from cache)